lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -989,16 +1001,33 @@ defmodule Mix.Compilers.Elixir do
     File.rm(manifest <> ".checkpoint")
   end
 
-  defp unless_previous_warnings_as_errors(previous_warnings, opts, {status, all_warnings}) do
+  defp unless_previous_warnings_as_errors(
+         previous_warnings,
+         opts,
+         {status, all_warnings, modules_diff}
+       ) do
     if previous_warnings != [] and opts[:warnings_as_errors] do
       message = "Compilation failed due to warnings while using the --warnings-as-errors option"
       IO.puts(:stderr, message)
-      {:error, all_warnings}
+      {:error, all_warnings, modules_diff}
     else
-      {status, all_warnings}
+      {status, all_warnings, modules_diff}
     end
   end
 
+  defp modules_diff(compiled_modules, pending_modules, all_modules, timestamp) do
+    compiled_modules_keys = Map.keys(compiled_modules)
+    pending_modules_keys = Map.keys(pending_modules)
+    all_modules_keys = Map.keys(all_modules)
+
+    %{
+      added: compiled_modules_keys -- all_modules_keys,
+      changed: Map.keys(Map.intersect(compiled_modules, all_modules)),
+      removed: (all_modules_keys -- compiled_modules_keys) -- pending_modules_keys,
+      timestamp: timestamp
+    }
```

Comment:
I am worried about this code being expensive for really large applications (20k modules). In such cases, we are changing 1-100 files, but we now need to traverse 20k to compute these statistics.

We could simplify the stats: instead of sending added/changed/removed, we just send "changed", noting that it implies added/removed/changed, and the subscriber can then lookup in disk the actual status. For the language server case, this means we may look up some modules that have been removed, but that's fine. They are the minority in most cases. For IEx/Phoenix, they don't care, they will just purge all.

If we don't want a single field, we can also do with `changed/removed`. But I think a single field would totally work on this case. WDYT?

Reply:
I've just pushed an update to never traverse all the modules. If you still prefer to have a single field, let me know :)

Reply:
This looks fantastic, I believe we can ship it. Something we can also try to do is to pass a function to PubSub. If someone is listening, we compute the event and send it. If nobody is listening, we don't. WDYT?

Reply:
If we do that, the function would close over `all_modules` and we would end up copying possibly huge amounts of data when passing it to the process, which may be worse than before, no?

Reply:
We should not be copying them at all? Couldn't we get open socket instances in the GenServer client and write to them directly?

Reply:
Oh I confused for the subscriber side. For publishing we only list the directory, so yeah we can make the message lazy. It will be a part of `Mix.Task.Compiler.notify_modules_compiled` API, but that should be fine. See b824a5d5b207a6803e5c7e3fb203f69ac685af09.